### PR TITLE
Zero and inf fix

### PR DESCRIPTION
**Pull request by sergeyf**
Re: #3 

@iskandr Please check it out to make sure these changes don't fundamentally break something else.

### CODE REVIEW

**Comment by amueller** [comment]
Maybe add a minimal regression test using
```python
X = np.array([[1, 1, np.NaN], [1, 1, 1]])
knnimpute.knn_impute_few_observed(X, np.isnan(X), k=1)
```
?

**Comment by sergeyf** [comment]
Yup, good idea. It's in there.

File: knnimpute/common.py [hunk 1]
```diff
@@ -30,8 +34,10 @@ def knn_initialize(X, missing_mask, verbose=False):
         # to put NaN's back in the data matrix for the distances function
         X_row_major[missing_mask] = np.nan
     D = all_pairs_normalized_distances(X_row_major, verbose=verbose)
-    # set diagonal of distance matrix to infinity since we don't want
+    # set diagonal of distance matrix to a large value since we don't want
     # points considering themselves as neighbors
-    for i in range(X.shape[0]):
-        D[i, i] = np.inf
+    max_dist = 1e6 * np.maximum(1, D.max())
```
Comment (iskandr) [inline knnimpute/common.py]:
```
In [1]: np.array([1,2,np.inf]).max()
Out[1]: inf
```

File: knnimpute/common.py [hunk 2]
```diff
@@ -30,7 +34,10 @@ def knn_initialize(X, missing_mask, verbose=False):
         # to put NaN's back in the data matrix for the distances function
         X_row_major[missing_mask] = np.nan
     D = all_pairs_normalized_distances(X_row_major, verbose=verbose)
-    # set diagonal of distance matrix to infinity since we don't want
+    # set diagonal of distance matrix to a large value since we don't want
     # points considering themselves as neighbors
-    np.fill_diagonal(D, np.inf)
+    max_dist = 1e6 * np.maximum(1, D.max())
```
Comment (iskandr) [inline knnimpute/common.py]: If infinity is in the distance matrix, won't this make `max_dist` infinity? 
```
In [1]: np.array([1,2,np.inf]).max()
Out[1]: inf
```

File: knnimpute/common.py [hunk 3]
```diff
@@ -19,7 +19,11 @@
 from .normalized_distance import all_pairs_normalized_distances
 
 
-def knn_initialize(X, missing_mask, verbose=False):
+def knn_initialize(X, 
+                   missing_mask, 
+                   verbose=False, 
+                   min_dist=1e-6, 
+                   max_dist_multiplier=1e6):
```
Comment (iskandr) [inline knnimpute/common.py:26]: What does `max_dist_multiplier` do?

File: knnimpute/common.py (hunk 1 above)
Comment (sergeyf) [reply knnimpute/common.py]: Oh right, I should do `D[np.isfinite(D)].max()`

Comment (sergeyf) [reply knnimpute/common.py]: Err unless that makes D empty. That's impossible, right?

File: knnimpute/common.py (hunk 3 above)
Comment (sergeyf) [reply knnimpute/common.py:26]: Oh right, it's supposed to be in the code instead of 1e6.

File: knnimpute/common.py (hunk 1 above)
Comment (iskandr) [reply knnimpute/common.py]: Well, I guess you could play it very safe and have:

```
finite_distances = D[np.isfinite(D)]
if len(finite_distances) > 0:
   max_dist = 1e6 * max(1, finite_distances.max())
else:
   max_dist = 1e6
```

**Comment by iskandr** [comment]
FYI, this is going to break the `few_observed_entries` implementation, which depends on being able to tell which distances are infinite due to no overlap between samples. 

```
    effective_infinity = 10 ** 6 * D[finite_distance_distance_mask].max()
    D[~finite_distance_distance_mask] = effective_infinity
    D_sorted = np.argsort(D, axis=1)
    inv_D = 1.0 / D
    D_valid_mask = D < effective_infinity
```

**Comment by iskandr** [comment]
One way to fix this for `few_observed_entries` would be to have `knn_initialize` return `max_dist`. I can do that as a second PR if you'd like.

**Comment by sergeyf** [comment]
@iskandr I think I fixed that a few commits back?

```
    effective_infinity = D[0, 0] # since diagonal was replaced by max_dist
    D_valid_mask = D < effective_infinity
    valid_distances_per_row = D_valid_mask.sum(axis=1)
```

**Comment by iskandr** [comment]
Didn't see that, looks good.

**Comment by iskandr** [comment]
Test failure: ```NameError: global name 'knnimpute' is not defined```

**Comment by sergeyf** [comment]
That's weird. I fixed that:

```
+def test_knn_minimal():
 +    X = np.array([[1, 1, np.NaN], [1, 1, 1]])
 +    res = knn_impute_few_observed(X, np.isnan(X), k=1)
 +    assert np.isnan(res).any() == False, \
 +        "Basic example did not get imputed: %s" % res 
```

**Comment by sergeyf** [comment]
Did you get the latest version?

**Comment by iskandr** [comment]
I took a peek at Travis but I guess it's behind on builds. Wonder why it's so slow today.

**Comment by iskandr** [comment]
Not sure which library this bubbles up from, but we get a warning on Python 2.7: 

```
Problem importing module variables.py: No module named backports.functools_lru_cache
```

Possibly fixed by depending on https://pypi.python.org/pypi/backports.functools_lru_cache/1.3? I can look into it later.

**Pull request merged**